Switch Travis CI to Xenial

chrisbra opened the pull request.

chrisbra committed:
Switch Travis CI to Xenial

Changes:
- Change how xvfb is called
  https://docs.travis-ci.com/user/gui-and-headless-browsers/
- install libtool-bin for comiling libvterm
- linking fails for libvterm when using clang and --coverage,
  so disable clang compilation for test_libvterm:

  This seems to be an error, compiling libvterm using clang with coverage
  https://travis-ci.org/vim/vim/jobs/519118820

```
libtool --quiet --mode=link --tag=CC clang -rpath /usr/local/lib -version-info 0:0:0 -o libvterm.la src/encoding.lo src/keyboard.lo src/mouse.lo src/parser.lo src/pen.lo src/state.lo src/termscreen.lo src/unicode.lo src/vterm.lo -L. -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,-E  --coverage -L/usr/local/lib -Wl,--as-needed
ar: `u' modifier ignored since `D' is the default (see `U')
libtool --quiet --mode=compile --tag=CC clang --coverage -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -Wall -Iinclude -std=c99 -Wpedantic -DINLINE="" -o t/harness.lo -c t/harness.c
libtool --quiet --mode=link --tag=CC clang --coverage -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -Wall -Iinclude -std=c99 -Wpedantic -DINLINE="" -o t/harness t/harness.lo libvterm.la -L. -Wl,-Bsymbolic-functions -Wl,-z,relro -Wl,-z,now -fstack-protector -rdynamic -Wl,-export-dynamic -Wl,-E  --coverage -L/usr/local/lib -Wl,--as-needed
/usr/bin/ld: t/.libs/harness: hidden symbol `llvm_gcda_end_file' in /usr/local/clang-7.0.0/lib/clang/7.0.0/lib/linux/libclang_rt.profile-x86_64.a(GCDAProfiling.c.o) is referenced by DSO
/usr/bin/ld: final link failed: Bad value
clang-7: error: linker command failed with exit code 1 (use -v to see invocation)
Makefile:78: recipe for target 't/harness' failed
make[2]: *** [t/harness] Error 1
```

That looks like an error on libclang side, so disable it for now.

Finally test86 fails when compiled with gcc on linux using
``
  -g -O1 -DABORT_ON_INTERNAL_ERROR -DEXITFREE -fsanitize=address -fno-omit-frame-pointer
``

The test fails and quite a bit of leaks are detected when compiled
against libpython2.7

Also noteworthy, it throws an error when running asan_symbolize:

```
/opt/pyenv/libexec/pyenv: line 43: cd: asan_symbolize-3.8: Not a directory
```

which looks in fact like this bug: pyenv/pyenv#807

mgedmin commented on .travis.yml:
```diff
@@ -69,14 +72,10 @@ branches:
-      - gettext
       - libperl-dev
       - python-dev
       - python3-dev
```
You probably want to add `libtool-bin` here.

chrisbra replied:
Interesting, my xenial vm image wanted libtool and not libtool-bin.

mgedmin replied:
/usr/bin/libtool on xenial comes from the [libtool-bin](https://packages.ubuntu.com/xenial/amd64/libtool-bin/filelist) package; the [libtool](https://packages.ubuntu.com/xenial/all/libtool/filelist) package only has /usr/bin/libtoolize.

(I don't know why.  It looks like trusty did not even have a libtool-bin package.)

chrisbra commented:
should work now, except for one build-failure:
https://travis-ci.org/vim/vim/jobs/519200082

```
VIMRUNTIME=../../runtime; export VIMRUNTIME;  ../vim -f  -u unix.vim -U NONE --noplugin --not-a-term -s dotest.in test86.in --cmd 'au SwapExists * let v:swapchoice = "e"' > /dev/null
Makefile:97: recipe for target 'test86.out' failed
make[2]: *** [test86.out] Error 23
make[2]: Leaving directory '/home/travis/build/vim/vim/src/testdir'
Makefile:2122: recipe for target 'scripttests' failed
make[1]: *** [scripttests] Error 2
make[1]: Leaving directory '/home/travis/build/vim/vim/src'
Makefile:36: recipe for target 'test' failed
make: *** [test] Error 2
```

So test86 breaks, but don't know why and then asan_symbolize also breaks when detecting the leaks. Don't know what to do against it...

brammool commented:
test86 is about Python.  You can perhaps check the Python version being used.  If you can't reproduce it, you can disable redirecting to /dev/null to see why the test fails.

chrisbra committed:
Update leak-suppress for Xenial

Ignore leaks for the following libraries
libperl.so
libpython2.7.so
libruby-2.3.so
libfontconfig
libc

brammool commented:
It looks like the test runs until the end, but doesn't get to the place where the output is compared.
Hmm, perhaps the problem is the memory leak?  We can suppress the errors, such as 
leak:PyDict_GetItemString

dpelle commented:
I see that CI failed because lsan found leaks in Python, Perl and Ruby libs.
We could blacklist those leaks in `vim/src/testdir/lsan-suppress.txt`.

Not sure if this works, but it might be something like this:
```
$ git diff ./src/testdir/lsan-suppress.txt
diff --git a/src/testdir/lsan-suppress.txt b/src/testdir/lsan-suppress.txt
index 5166fbf5c..e2a4d2e1a 100644
--- a/src/testdir/lsan-suppress.txt
+++ b/src/testdir/lsan-suppress.txt
@@ -1,3 +1,6 @@
-# Suppress leaks from X libraries on Ubuntu trusty.
+# Suppress leaks from X libraries on Ubuntu xenial.
+leak:libperl.so
+leak:libpython2.7.so
+leak:libruby-2.3.so
 leak:libX11.so.6
 leak:libXt.so.6
 leakak:libXt.so.6
```

chrisbra commented:
Thanks, I'll try the updated leak-suppress.

chrisbra commented:
I have tested several different changes. I have now cleaned up the commits a bit. While travis is [still running](https://travis-ci.org/vim/vim/builds/520326078?utm_source=github_status&utm_medium=notification) let me document the changes done so far. (It will probably fail for the test with `SANITIZER_CFLAGS="-g -O1 -DABORT_ON_INTERNAL_ERROR -DEXITFREE...`, but see below).

There are now mainly two changes:

- switching to xenial in .travis.yml
  Also, since xenial comes with Xvfb installed, we can simply use
```
    services:
      - xvfb
```
to turn on the virtual frame buffer

- I disabled test_libvterm for clang, because it caused problems with linking. The linker complained about a hidden symbol in libclang. I suppose that is a compiler bug, but it happens to make the specific test run fail, so the simplest solution was to disable it.

- I added a couple of more libraries to the leak-suppression config file. This is, because some tests when run under asan/lsan report leaks in libpython, libruby, libperl, libfontconfig and even in libc. This is not really useful, so I added them to the `src/testdir/lsan-suppress.txt` file

Still I see the following test failure, when run under asan:
```
==14783==ERROR: AddressSanitizer: stack-buffer-underflow on address 0x7ffc43cc0cc0 at pc 0x7f9919b2dbec bp 0x7ffc43cc09a0 sp 0x7ffc43cc0148
WRITE of size 104 at 0x7ffc43cc0cc0 thread T0
    #0 0x7f9919b2dbeb in __asan_memset (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x8cbeb)
    #1 0x4a2677 in memset /usr/include/x86_64-linux-gnu/bits/string3.h:90
    #2 0x4a2677 in get_lval /home/travis/build/vim/vim/src/eval.c:1933
    #3 0x7cc6bf in trans_function_name /home/travis/build/vim/vim/src/userfunc.c:1703
    #4 0x7d655d in ex_call /home/travis/build/vim/vim/src/userfunc.c:3154
    #5 0x52c4b3 in do_one_cmd /home/travis/build/vim/vim/src/ex_docmd.c:2526
    #6 0x52c4b3 in do_cmdline /home/travis/build/vim/vim/src/ex_docmd.c:1033
    #7 0x52fb6b in do_cmdline_cmd /home/travis/build/vim/vim/src/ex_docmd.c:634
    #8 0x49ca6d in assert_fails /home/travis/build/vim/vim/src/eval.c:9592
    #9 0x4d87f1 in f_assert_fails /home/travis/build/vim/vim/src/evalfunc.c:1649
    #10 0x4dc0d4 in call_internal_func /home/travis/build/vim/vim/src/evalfunc.c:1142
    #11 0x7d57e9 in call_func /home/travis/build/vim/vim/src/userfunc.c:1552
    #12 0x7d5df2 in get_func_tv /home/travis/build/vim/vim/src/userfunc.c:453
    #13 0x7d69cf in ex_call /home/travis/build/vim/vim/src/userfunc.c:3216
    #14 0x52c4b3 in do_one_cmd /home/travis/build/vim/vim/src/ex_docmd.c:2526
    #15 0x52c4b3 in do_cmdline /home/travis/build/vim/vim/src/ex_docmd.c:1033
    #16 0x7d4d84 in call_user_func /home/travis/build/vim/vim/src/userfunc.c:999
    #17 0x7d4d84 in call_func /home/travis/build/vim/vim/src/userfunc.c:1533
    #18 0x7d5df2 in get_func_tv /home/travis/build/vim/vim/src/userfunc.c:453
    #19 0x7d69cf in ex_call /home/travis/build/vim/vim/src/userfunc.c:3216
    #20 0x52c4b3 in do_one_cmd /home/travis/build/vim/vim/src/ex_docmd.c:2526
    #21 0x52c4b3 in do_cmdline /home/travis/build/vim/vim/src/ex_docmd.c:1033
    #22 0x49f5bb in ex_execute /home/travis/build/vim/vim/src/eval.c:8644
    #23 0x52c4b3 in do_one_cmd /home/travis/build/vim/vim/src/ex_docmd.c:2526
    #24 0x52c4b3 in do_cmdline /home/travis/build/vim/vim/src/ex_docmd.c:1033
    #25 0x7d4d84 in call_user_func /home/travis/build/vim/vim/src/userfunc.c:999
    #26 0x7d4d84 in call_func /home/travis/build/vim/vim/src/userfunc.c:1533
    #27 0x7d5df2 in get_func_tv /home/travis/build/vim/vim/src/userfunc.c:453
    #28 0x7d69cf in ex_call /home/travis/build/vim/vim/src/userfunc.c:3216
    #29 0x52c4b3 in do_one_cmd /home/travis/build/vim/vim/src/ex_docmd.c:2526
    #30 0x52c4b3 in do_cmdline /home/travis/build/vim/vim/src/ex_docmd.c:1033
    #31 0x50a439 in do_source /home/travis/build/vim/vim/src/ex_cmds2.c:4607
    #32 0x50b8d8 in cmd_source /home/travis/build/vim/vim/src/ex_cmds2.c:4227
    #33 0x50ba5c in ex_source /home/travis/build/vim/vim/src/ex_cmds2.c:4202
    #34 0x52c4b3 in do_one_cmd /home/travis/build/vim/vim/src/ex_docmd.c:2526
    #35 0x52c4b3 in do_cmdline /home/travis/build/vim/vim/src/ex_docmd.c:1033
    #36 0x52fb6b in do_cmdline_cmd /home/travis/build/vim/vim/src/ex_docmd.c:634
    #37 0x8a2946 in exe_commands /home/travis/build/vim/vim/src/main.c:2940
    #38 0x8a2946 in vim_main2 /home/travis/build/vim/vim/src/main.c:795
    #39 0x8a5cbe in main /home/travis/build/vim/vim/src/main.c:438
    #40 0x7f99159bb82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #41 0x441ed8 in _start (/home/travis/build/vim/vim/src/vim+0x441ed8)

Address 0x7ffc43cc0cc0 is located in stack of thread T0 at offset 0 in frame
    #0 0x8342fc in luaV_list /home/travis/build/vim/vim/src/if_lua.c:1678

  This frame has 1 object(s):
    [32, 48) 'v' <== Memory access at offset 0 partially underflows this variable
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-underflow ??:0 __asan_memset
Shadow bytes around the buggy address:
  0x100008790140: 00 00 00 00 f1 f1 f1 f1 00 f4 f4 f4 f2 f2 f2 f2
  0x100008790150: 00 f4 f4 f4 f2 f2 f2 f2 00 f4 f4 f4 f2 f2 f2 f2
  0x100008790160: 00 f4 f4 f4 f2 f2 f2 f2 00 00 f4 f4 f2 f2 f2 f2
  0x100008790170: 00 00 f4 f4 f3 f3 f3 f3 00 00 00 00 00 00 00 00
  0x100008790180: 00 00 00 00 00 00 00 00 f1 f1 f1 f1 04 f4 f4 f4
=>0x100008790190: f2 f2 f2 f2 00 00 00 00[f1]f1 f1 f1 00 00 f4 f4
  0x1000087901a0: 00 f4 f4 f4 f2 f2 f2 f2 00 00 04 f4 f3 f3 f3 f3
  0x1000087901b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1000087901c0: 00 00 00 00 00 00 f1 f1 f1 f1 04 f4 f4 f4 f2 f2
  0x1000087901d0: f2 f2 04 f4 f4 f4 f2 f2 f2 f2 00 f4 f4 f4 f2 f2
  0x1000087901e0: f2 f2 00 f4 f4 f4 f2 f2 f2 f2 00 00 f4 f4 f2 f2
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
==14783==ABORTING
```
This happens in test_lua.vim. I have no idea if this is a false positive or not and how to fix this one.

Also noteworthy, there is an annoying output when running `asan-symbolize`:

    /opt/pyenv/libexec/pyenv: line 43: cd: asan_symbolize-3.8: Not a directory

That makes asan-symbolize not output anything useful, except the leak-suppresion info. Don't know how to fix this one, as this does not happen in my local xenial copy. Also I think the following could be disabled in [.travis.yml](https://github.com/vim/vim/blob/8dc4c7292364d31cfb8afa8a67984b78b657b192/.travis.yml#L162)

    [[ -z "${log}" ]] # exit 1 if there are ASAN logs

I think we should not abort too early, just because there are asan errors, let the tests continue in any case.

brammool commented:
Christian wrote:

> I have tested several different changes. I have now cleaned up the
> commits a bit. While travis is [still
> running](https://travis-ci.org/vim/vim/builds/520326078?utm_source=github_status&utm_medium=notification)
> let me document the changes done so far. (It will probably fail for
> the test with `SANITIZER_CFLAGS="-g -O1 -DABORT_ON_INTERNAL_ERROR
> -DEXITFREE...`, but see below).
> 
> There are now mainly two changes:
> 
> - switching to xenial in .travis.yml
>   Also, since xenial comes with Xvfb installed, we can simply use
> ```
>     services:
>       - xvfb
> ```
> to turn on the virtual frame buffer
> 
> - I disabled test_libvterm for clang, because it caused problems with
> linking. The linker complained about a hidden symbol in libclang. I
> suppose that is a compiler bug, but it happens to make the specific
> test run fail, so the simplest solution was to disable it.

Strange.  Clang usually works well.

> - I added a couple of more libraries to the leak-suppression config
> file. This is, because some tests when run under asan/lsan report
> leaks in libpython, libruby, libperl, libfontconfig and even in libc.
> This is not really useful, so I added them to the
> `src/testdir/lsan-suppress.txt` file

That appears to have worked.

> Still I see the following test failure, when run under asan:
> ```
> ==14783==ERROR: AddressSanitizer: stack-buffer-underflow on address 0x7ffc43cc0cc0 at pc 0x7f9919b2dbec bp 0x7ffc43cc09a0 sp 0x7ffc43cc0148
> WRITE of size 104 at 0x7ffc43cc0cc0 thread T0
>     #0 0x7f9919b2dbeb in __asan_memset (/usr/lib/x86_64-linux-gnu/libasan.so.2+0x8cbeb)
>     #1 0x4a2677 in memset /usr/include/x86_64-linux-gnu/bits/string3.h:90
>     #2 0x4a2677 in get_lval /home/travis/build/vim/vim/src/eval.c:1933
>     #3 0x7cc6bf in trans_function_name /home/travis/build/vim/vim/src/userfunc.c:1703
>     #4 0x7d655d in ex_call /home/travis/build/vim/vim/src/userfunc.c:3154
>     #5 0x52c4b3 in do_one_cmd /home/travis/build/vim/vim/src/ex_docmd.c:2526
>     #6 0x52c4b3 in do_cmdline /home/travis/build/vim/vim/src/ex_docmd.c:1033
>     #7 0x52fb6b in do_cmdline_cmd /home/travis/build/vim/vim/src/ex_docmd.c:634
>     #8 0x49ca6d in assert_fails /home/travis/build/vim/vim/src/eval.c:9592
>     #9 0x4d87f1 in f_assert_fails /home/travis/build/vim/vim/src/evalfunc.c:1649
>     #10 0x4dc0d4 in call_internal_func /home/travis/build/vim/vim/src/evalfunc.c:1142
>     #11 0x7d57e9 in call_func /home/travis/build/vim/vim/src/userfunc.c:1552
>     #12 0x7d5df2 in get_func_tv /home/travis/build/vim/vim/src/userfunc.c:453
>     #13 0x7d69cf in ex_call /home/travis/build/vim/vim/src/userfunc.c:3216
>     #14 0x52c4b3 in do_one_cmd /home/travis/build/vim/vim/src/ex_docmd.c:2526
>     #15 0x52c4b3 in do_cmdline /home/travis/build/vim/vim/src/ex_docmd.c:1033
>     #16 0x7d4d84 in call_user_func /home/travis/build/vim/vim/src/userfunc.c:999
>     #17 0x7d4d84 in call_func /home/travis/build/vim/vim/src/userfunc.c:1533
>     #18 0x7d5df2 in get_func_tv /home/travis/build/vim/vim/src/userfunc.c:453
>     #19 0x7d69cf in ex_call /home/travis/build/vim/vim/src/userfunc.c:3216
>     #20 0x52c4b3 in do_one_cmd /home/travis/build/vim/vim/src/ex_docmd.c:2526
>     #21 0x52c4b3 in do_cmdline /home/travis/build/vim/vim/src/ex_docmd.c:1033
>     #22 0x49f5bb in ex_execute /home/travis/build/vim/vim/src/eval.c:8644
>     #23 0x52c4b3 in do_one_cmd /home/travis/build/vim/vim/src/ex_docmd.c:2526
>     #24 0x52c4b3 in do_cmdline /home/travis/build/vim/vim/src/ex_docmd.c:1033
>     #25 0x7d4d84 in call_user_func /home/travis/build/vim/vim/src/userfunc.c:999
>     #26 0x7d4d84 in call_func /home/travis/build/vim/vim/src/userfunc.c:1533
>     #27 0x7d5df2 in get_func_tv /home/travis/build/vim/vim/src/userfunc.c:453
>     #28 0x7d69cf in ex_call /home/travis/build/vim/vim/src/userfunc.c:3216
>     #29 0x52c4b3 in do_one_cmd /home/travis/build/vim/vim/src/ex_docmd.c:2526
>     #30 0x52c4b3 in do_cmdline /home/travis/build/vim/vim/src/ex_docmd.c:1033
>     #31 0x50a439 in do_source /home/travis/build/vim/vim/src/ex_cmds2.c:4607
>     #32 0x50b8d8 in cmd_source /home/travis/build/vim/vim/src/ex_cmds2.c:4227
>     #33 0x50ba5c in ex_source /home/travis/build/vim/vim/src/ex_cmds2.c:4202
>     #34 0x52c4b3 in do_one_cmd /home/travis/build/vim/vim/src/ex_docmd.c:2526
>     #35 0x52c4b3 in do_cmdline /home/travis/build/vim/vim/src/ex_docmd.c:1033
>     #36 0x52fb6b in do_cmdline_cmd /home/travis/build/vim/vim/src/ex_docmd.c:634
>     #37 0x8a2946 in exe_commands /home/travis/build/vim/vim/src/main.c:2940
>     #38 0x8a2946 in vim_main2 /home/travis/build/vim/vim/src/main.c:795
>     #39 0x8a5cbe in main /home/travis/build/vim/vim/src/main.c:438
>     #40 0x7f99159bb82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
>     #41 0x441ed8 in _start (/home/travis/build/vim/vim/src/vim+0x441ed8)
> 
> Address 0x7ffc43cc0cc0 is located in stack of thread T0 at offset 0 in frame
>     #0 0x8342fc in luaV_list /home/travis/build/vim/vim/src/if_lua.c:1678

This is bogus, since nothing in if_lua.c is in the stack backtrace.
 
> This happens in test_lua.vim. I have no idea if this is a false
> positive or not and how to fix this one.

Perhaps it is related to the call to LuaL_error() in the reported line?
The docs say the function Raises an error.  Perhaps it is caught
somewhere, and this causes the "false positive" that asan mentions?
 
> Also noteworthy, there is an annoying output when running `asan-symbolize`:
> 
>     /opt/pyenv/libexec/pyenv: line 43: cd: asan_symbolize-3.8: Not a directory
> 
> That makes asan-symbolize not output anything useful, except the leak-suppresion info. Don't know how to fix this one, as this does not happen in my local xenial copy. Also I think the following could be disabled in [.travis.yml](https://github.com/vim/vim/blob/8dc4c7292364d31cfb8afa8a67984b78b657b192/.travis.yml#L162)
> 
>     [[ -z "${log}" ]] # exit 1 if there are ASAN logs
> 
> I think we should not abort too early, just because there are asan
> errors, let the tests continue in any case.

Thanks for looking into all this!

-- 
ARTHUR: Right! Knights! Forward!
   ARTHUR leads a charge toward the castle.  Various shots of them battling on,
   despite being hit by a variety of farm animals.
                 "Monty Python and the Holy Grail" PYTHON (MONTY) PICTURES LTD

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

dpelle commented:
@brammool  wrote:

> > Address 0x7ffc43cc0cc0 is located in stack of thread T0 at offset 0 in frame
> >      #0 0x8342fc in luaV_list /home/travis/build/vim/vim/src/if_lua.c:1678
>
> This is bogus, since nothing in if_lua.c is in the stack backtrace.

It's not necessarily bogus.  The address of a variable in stack in if_lua.c could
be stored in a data structure and then accessed much later in a totally unrelated
stack. That said, I don't understand the bug here after looking at the code.

brammool commented:
Dominique wrote:

> @brammool  wrote:
> 
> > > Address 0x7ffc43cc0cc0 is located in stack of thread T0 at offset 0 in frame
> > >      #0 0x8342fc in luaV_list /home/travis/build/vim/vim/src/if_lua.c:1678
> >
> > This is bogus, since nothing in if_lua.c is in the stack backtrace.
> 
> It's not necessarily bogus.  The address of a variable in stack in
> if_lua.c could be stored in a data structure and then accessed much
> later in a totally unrelated stack. That said, I don't understand the
> bug here after looking at the code. 

The reported access violation happens in memset(), and I don't see how
that can be wrong, considering the stack trace.  It's like ASAN has
remembered this position in the stack from something that happened
before, and failed to remove it when the stack was unwinding.  That's
also mentioned in the "false positive" note.  So I was wondering if Lua
does something special with exceptions.

-- 
In his lifetime van Gogh painted 486 oil paintings. Oddly enough, 8975
of them are to be found in the United States.

 /// Bram Moolenaar -- Bram@Moolenaar.net -- http://www.Moolenaar.net   \\\
///        sponsor Vim, vote for features -- http://www.Vim.org/sponsor/ \\\
\\\  an exciting new programming language -- http://www.Zimbu.org        ///
 \\\            help me help AIDS victims -- http://ICCF-Holland.org    ///

chrisbra committed:
blacklist get_lval for ASAN

chrisbra commented:
I also wondered about the stack-underflow in `memset`. So I considered to blacklist `get_lval()` function for asan like described [here](https://github.com/google/sanitizers/wiki/AddressSanitizer#turning-off-instrumentation). I am not super comfortable with blacklisting functions from vim, but I tried to see if this at least worked:

```patch
diff --git a/src/eval.c b/src/eval.c
index 5f45d69..e2ad07f 100644
--- a/src/eval.c
+++ b/src/eval.c
@@ -11,6 +11,11 @@
  * eval.c: Expression evaluation.
  */
 #define USING_FLOAT_STUFF
+#if defined(__clang__) || defined (__GNUC__)
+# define ATTRIBUTE_NO_SANITIZE_ADDRESS __attribute__((no_sanitize_address))
+#else
+# define ATTRIBUTE_NO_SANITIZE_ADDRESS
+#endif

 #include "vim.h"

@@ -1906,7 +1911,8 @@ ex_let_one(
  * When an evaluation error occurs "lp->ll_name" is NULL;
  * Returns NULL for a parsing error.  Still need to free items in "lp"!
  */
-    char_u *
+ATTRIBUTE_NO_SANITIZE_ADDRESS
+       char_u *
 get_lval(
     char_u     *name,
     typval_T   *rettv,
```

However the build now [fails](https://travis-ci.org/vim/vim/jobs/521101902) with:
```
In file included from /usr/include/features.h:367:0,
                 from /usr/include/limits.h:25,
                 from /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed/limits.h:168,
                 from /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed/syslimits.h:7,
                 from /usr/lib/gcc/x86_64-linux-gnu/5/include-fixed/limits.h:34,
                 from vim.h:58,
                 from eval.c:15:
eval.c: In function ‘get_lval’:
/usr/include/x86_64-linux-gnu/bits/string3.h:78:1: error: inlining failed in call to always_inline ‘memset’: function attribute mismatch
 __NTH (memset (void *__dest, int __ch, size_t __len))
 ^
In file included from eval.c:15:0:
vim.h:1702:37: error: called from here
 # define vim_memset(ptr, c, size)   memset((ptr), (c), (size))
                                     ^
eval.c:1934:5: note: in expansion of macro ‘vim_memset’
     vim_memset(lp, 0, sizeof(lval_T));
     ^
Makefile:2980: recipe for target 'objects/eval.o' failed
make[2]: *** [objects/eval.o] Error 1
make[2]: Leaving directory '/home/travis/build/vim/vim/src'
```

Any ideas?

chrisbra commented:
if no-one has and idea, I guess we need to skip switching to xenial.
